### PR TITLE
Issue #3007303 by vnech: Cross-posting nodes in multiple groups

### PR DESCRIPTION
**Pull request by nechai**
## Problem
There is no possibility to add one node to multiple groups.

## Solution
- Change field "groups" cardinality to unlimited;
- Add ability for SM to chose what content types can be posted in multiple groups;
- Add ability to CM+ to post content in multiple groups;

## Issue tracker
- https://www.drupal.org/project/social/issues/3007303
- https://getopensocial.atlassian.net/browse/YANG-5092
- https://getopensocial.atlassian.net/browse/YANG-5020

## How to test
- [x] As SM go to _/admin/config/opensocial/social-group_ and enable cross-posting feature
- [x] As SM+ go to add/edit entity type you enabled in settings and in field "Groups" chose multiple groups
- [x] Check if the entity is added as content to groups you choose in field "Groups"

## Release notes
Add ability or CM+ to cross-post content in multiple groups.

## Change Record
- New message templates:
  - `create_event_gc`
  - `create_topic_gc`
- Changed message templates:
  - `create_content_in_joined_group`
  - `create_event_group`
  - `create_topic_group`
  - `moved_content_between_groups`
- New plugins:
  - `modules/social_features/social_activity/src/Plugin/ActivityEntityCondition/GroupContentSingleActivityEntityCondition.php`
- Changed plugins:
  - `modules/social_features/social_group/src/Plugin/Field/FieldWidget/SocialGroupSelectorWidget.php`
  - `modules/custom/activity_basics/src/Plugin/ActivityContext/GroupActivityContext.php`
  - `modules/custom/activity_viewer/src/Plugin/views/filter/ActivityPostVisibilityAccess.php`
- New tokens:
  - `[message:count_groups_per_node]`
- New hook:
  - `hook_social_group_cross_posting_alter(array &$content_types)`
- New service:
  - `social_group.cross_posting`
- Refactoring:
  -  `modules/custom/entity_access_by_field/src/EntityAccessHelper.php`

### CODE REVIEW

File: modules/social_features/social_group/src/Form/SocialGroupSettings.php
Comment (ribel) [inline modules/social_features/social_group/src/Form/SocialGroupSettings.php:248]: @nechai We usually start to alter hook names with module name.
So maybe we can name it `social_group_cross_posting`
Also, an example needs to be added to the `social_group.api.php` file.

**Review by ribel — CHANGES_REQUESTED** [review]
@nechai Please resolve conflict and rename alter hook.

File: modules/social_features/social_group/src/Form/SocialGroupSettings.php
Comment (nechai) [reply modules/social_features/social_group/src/Form/SocialGroupSettings.php]: @ribel changed hook name according to your suggestions

File: modules/social_features/social_group/src/Form/SocialGroupSettings.php
Comment (ribel) [inline modules/social_features/social_group/src/Form/SocialGroupSettings.php:149]: @nechai I think it would be clearer to rename this label and other places in code it to `Content types` since we implement cross-posting only for content types, but not for other entities Posts.

**Review by ribel — APPROVED** [review]

File: modules/social_features/social_activity/social_activity.tokens.inc
Comment (ribel) [inline modules/social_features/social_activity/social_activity.tokens.inc:29]: Rename `$message` to `$tokens` to be consistent with future changes here: #2277

File: modules/social_features/social_activity/social_activity.tokens.inc
Comment (ribel) [inline modules/social_features/social_activity/social_activity.tokens.inc:130]: Wtap word `more` in `t()` function as well.

File: modules/social_features/social_activity/src/Plugin/ActivityEntityCondition/GroupContentSingleActivityEntityCondition.php
Comment (ribel) [inline modules/social_features/social_activity/src/Plugin/ActivityEntityCondition/GroupContentSingleActivityEntityCondition.php:6]: Remove unused classes:
```
use Drupal\Core\Database\Connection;
use Drupal\group\Entity\GroupContentInterface;
use Drupal\group\Plugin\GroupContentEnablerManagerInterface;
use function MongoDB\Driver\Monitoring\removeSubscriber;
```

File: modules/social_features/social_activity/src/Plugin/ActivityEntityCondition/GroupContentMultipleActivityEntityCondition.php
Comment (ribel) [inline modules/social_features/social_activity/src/Plugin/ActivityEntityCondition/GroupContentMultipleActivityEntityCondition.php:6]: Remove unused classes:
```
use Drupal\Core\Database\Connection;
use Drupal\group\Entity\GroupContentInterface;
use Drupal\group\Plugin\GroupContentEnablerManagerInterface;
use function MongoDB\Driver\Monitoring\removeSubscriber;
```

File: modules/social_features/social_group/config/install/social_group.settings.yml
Comment (ribel) [inline modules/social_features/social_group/config/install/social_group.settings.yml:8]: To be consistent and inclusive I would propose use just `group_types` variable name, title and logic in all places.

File: modules/social_features/social_group/social_group.module
Comment (ribel) [inline modules/social_features/social_group/social_group.module:1282]: To be consistent and inclusive I would propose use just group_types variable name, title and logic in all places.

File: modules/social_features/social_group/src/Form/SocialGroupSettings.php
Comment (ribel) [inline modules/social_features/social_group/src/Form/SocialGroupSettings.php:160]: To be consistent and inclusive I would propose use just group_types variable name, title and logic in all places.

File: modules/social_features/social_group/src/Form/SocialGroupSettings.php
Comment (ribel) [inline modules/social_features/social_group/src/Form/SocialGroupSettings.php:210]: To be consistent and inclusive I would propose use just group_types variable name, title and logic in all places.

**Review by ribel — CHANGES_REQUESTED** [review]
@nechai Please check my comments after review.

**Comment by ribel** [comment]
@nechai Also since we introduce quite a big change here, please add more description about new changes to the Change Record section and create a record about new hook:

- New message templates
- Changed message templates
- New plugins
- New tokens
- Changed views
- Other major changes in the Activity stream architecture

File: modules/custom/activity_basics/src/Plugin/ActivityContext/GroupActivityContext.php
Comment (ribel) [reply modules/custom/activity_basics/src/Plugin/ActivityContext/GroupActivityContext.php:108]: @nechai Question is still relevant :)

**Comment by nechai** [comment]
Hi @ronaldtebrake!
I have a question regarding configuration updating. 

Is it required to place configs to /config/static folder? Can I replace only configs fields, not a whole config ([as like here (see last hook update)](https://github.com/goalgorilla/open_social/blob/cfc3af52de75012f8e3eef72d052ad4f33ef5eb2/modules/social_features/social_mentions/social_mentions.install)) as it can break config (remove `uuid` and `_core` fields)?

**Comment by ronaldtebrake** [comment]
Hi @nechai 

There are many ways for you to update your configuration using update hooks. It isn't required to use `config/static`

But loading config from `config/install` is never a good idea, because it can change. You want a static representation of the update at that moment in time. So using `config/static` is fine. Using the update helper module, which helps you only changing what you really want to change is also fine.

Or you can always just write your own update hook. For example like: https://github.com/goalgorilla/open_social/blob/ec061b5d073406d23a693a99bd936f6207221df9/modules/custom/social_advanced_queue/social_advanced_queue.install#L13

Does this help?

**Review by ribel — APPROVED** [review]

File: modules/social_features/social_group/src/Form/SocialGroupSettings.php
Comment (ribel) [inline modules/social_features/social_group/src/Form/SocialGroupSettings.php:162]: @nechai It's better to rename title and logic of this variable from `Disable` to `Enable`
Also to be consistent you can just rename title to the `Group types` and use description fo the explanation.

File: modules/social_features/social_group/social_group.module
Comment (ribel) [inline modules/social_features/social_group/social_group.module:1212]: @nechai It's better to rename name and logic of this variable from `disabled` to `enabled`

**Review by ribel — CHANGES_REQUESTED** [review]
@nechai All looks good, but two things needs to be fixed before we can merge this PR:

1.  Rename title and logic of `group_types` variable from `Disable` to `Enable`
2. Fix behat tests, maybe you can investigate it locally with clean Open Social setup or create new PR if you can't reproduce it.

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2300-acboxzdye5wa8xo8f7a0iz6z01elmag8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/609cccbccde720d566666ab2

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2300-acboxzdye5wa8xo8f7a0iz6z01elmag8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/609cccbccde720d566666ab2

File: modules/custom/entity_access_by_field/src/EntityAccessHelper.php
Comment (ronaldtebrake) [inline modules/custom/entity_access_by_field/src/EntityAccessHelper.php:30]: Nice!

File: modules/social_features/social_activity/social_activity.install
Comment (ronaldtebrake) [inline modules/social_features/social_activity/social_activity.install:268]: Would be great if we can update this to 10 number range as we are probably not backporting this feature.

File: modules/social_features/social_activity/social_activity.tokens.inc
Comment (ronaldtebrake) [inline modules/social_features/social_activity/social_activity.tokens.inc:17]: //cc @navneet0693 have a look for our email notification PR, we need to make sure this token works when merging this back in our PR later.

File: modules/social_features/social_activity/social_activity.tokens.inc
Comment (ronaldtebrake) [inline modules/social_features/social_activity/social_activity.tokens.inc:53]: //cc @ronaldtebrake lets check if we need to update PHP Stan to ensure we don't get any issues when using 
`            $target_type = $message->field_message_related_object->target_type;`

File: modules/social_features/social_activity/social_activity.tokens.inc
Comment (ronaldtebrake) [inline modules/social_features/social_activity/social_activity.tokens.inc:131]: I think we can use the format Plural here in order to ensure we render the string based on the Group count
e.g. + 1 more vs +2/3/4 more etc. That should help our customers in their translation efforts.

File: modules/social_features/social_group/modules/social_group_flexible_group/social_group_flexible_group.install
Comment (ronaldtebrake) [inline modules/social_features/social_group/modules/social_group_flexible_group/social_group_flexible_group.install:405]: Lets make sure we use the 10 number range here for the update hook.

File: modules/social_features/social_group/modules/social_group_flexible_group/social_group_flexible_group.install
Comment (ronaldtebrake) [inline modules/social_features/social_group/modules/social_group_flexible_group/social_group_flexible_group.install:407]: For update hooks its better to load data from a static folder instead of the /config/install, as the content inside that can change before your update hook fires and you don't know the outcome. Keeping that consistent would be great! :D

File: modules/social_features/social_group/modules/social_group_secret/social_group_secret.install
Comment (ronaldtebrake) [inline modules/social_features/social_group/modules/social_group_secret/social_group_secret.install:129]: Same applies here :D just to make sure we don't lose track

File: modules/social_features/social_group/social_group.api.php
Comment (ronaldtebrake) [inline modules/social_features/social_group/social_group.api.php:204]: Nice addition 👍 🚀

File: modules/social_features/social_group/social_group.install
Comment (ronaldtebrake) [inline modules/social_features/social_group/social_group.install:676]: same applies here

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2300-acboxzdye5wa8xo8f7a0iz6z01elmag8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/609cccbccde720d566666ab2

**Comment by nechai** [comment]
Hi, @ronaldtebrake @ribel 
I added the requested changes after code review. Only one thing remains - when the user converts group type (for example, from "Flexible Group" to "Open Group") and if cross-group posting is disabled for this group type the user will not be able to edit the node belong to this group. Do you have any idea how we can fix this?🤔

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2300-acboxzdye5wa8xo8f7a0iz6z01elmag8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/609cccbccde720d566666ab2

**Comment by ronaldtebrake** [comment]
> Hi, @ronaldtebrake @ribel
> I added the requested changes after code review. Only one thing remains - when the user converts group type (for example, from "Flexible Group" to "Open Group") and if cross-group posting is disabled for this group type the user will not be able to edit the node belong to this group. Do you have any idea how we can fix this?🤔

Interesting use case!
Please check with our product owner as well from a vision aspect.
I'm inclined to make a decision as to not support changing group types of existing groups anymore as we will be migrating everything to flexible groups in the future. So you dont have to switch group types anymore, you can rather switch the settings of a flexible group. But that only works from 11.0.0 onwards (or whenever that work lands). But maybe we can make that decision as part of this feature as well.

**Comment by nechai** [comment]
> > Hi, @ronaldtebrake @ribel
> > I added the requested changes after code review. Only one thing remains - when the user converts group type (for example, from "Flexible Group" to "Open Group") and if cross-group posting is disabled for this group type the user will not be able to edit the node belong to this group. Do you have any idea how we can fix this?🤔
> 
> Interesting use case!
> Please check with our product owner as well from a vision aspect.
> I'm inclined to make a decision as to not support changing group types of existing groups anymore as we will be migrating everything to flexible groups in the future. So you dont have to switch group types anymore, you can rather switch the settings of a flexible group. But that only works from 11.0.0 onwards (or whenever that work lands). But maybe we can make that decision as part of this feature as well.

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2300-uwcurgu7juaads0pprjprkqww7u7dqrf.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6100012c74246816e9689d98

**Comment by ribel** [comment]
Hey @ronaldtebrake, here is what Moritz said:

> What Ronald said is correct. We can disregard that issue as we will have removed all other group types from the product by the end of Q3

So does that mean we can merge this feature since all other issues seems to be resolved?

**Comment by ronaldtebrake** [comment]
That does mean it can only be in 11.0.0, that is where we will remove all our other group types.. is that feasible for the deadline as well? If not we might need to have a quickfix in place for 10.3.0.
So for example you can't change a group type when there is content cross posted inside that group. Something like that makes it BC still. What do you think?

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2300-uwcurgu7juaads0pprjprkqww7u7dqrf.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6100012c74246816e9689d98

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2300-uwcurgu7juaads0pprjprkqww7u7dqrf.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6100012c74246816e9689d98

**Comment by nechai** [comment]
Hi @ronaldtebrake @ribel 
I added restriction regarding changing "group type" ([see commit](https://github.com/goalgorilla/open_social/pull/2300/commits/16d396321e0bccd480daa2a1004b586b4d336062)). 
Now, if the user wants to change group type, he will be able to do this only if "old group type" and "new group type" supports the cross-group posting feature.

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2300-uwcurgu7juaads0pprjprkqww7u7dqrf.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6100012c74246816e9689d98

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2300-uwcurgu7juaads0pprjprkqww7u7dqrf.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6100012c74246816e9689d98

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2300-uwcurgu7juaads0pprjprkqww7u7dqrf.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6100012c74246816e9689d98

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2300-uwcurgu7juaads0pprjprkqww7u7dqrf.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6100012c74246816e9689d98

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2300-uwcurgu7juaads0pprjprkqww7u7dqrf.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6100012c74246816e9689d98

**Comment by navneet0693** [comment]
Screenshot 2021-08-12 at 6 11 24 PM

@nechai There are few suggestions to change:

1. If possible, mark "Content-Type" and "Group type" as required, as it seems that it is mandatory to select both.
2. This will also remove confusion for site managers to just enable and then try cross-posting without "checking" the options in content type and group type.
3. All the descriptions are a little bit too technical from an end-user perspective.

**Comment by nechai** [comment]
Hi @navneet0693!
I made these two fields ("Content types" & "Group types") required.
Regarding descriptions, if you have some suggestions what should be changed we can add them here ☺️

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2300-uwcurgu7juaads0pprjprkqww7u7dqrf.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6100012c74246816e9689d98

**Comment by ribel** [comment]
@nechai I see two changed behat tests are failing even after re-run.
Also from now, we try to avoid new rules in `phpstan-baseline.neon` so please investigate how we can fix CS in other ways.

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2300-uwcurgu7juaads0pprjprkqww7u7dqrf.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6100012c74246816e9689d98

File: modules/social_features/social_activity/config/update/social_activity_update_10302.yml
Comment (dmytriikaiun) [inline modules/social_features/social_activity/config/update/social_activity_update_10302.yml:7]: @nechai should it be also added to the install configuration?

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2300-uwcurgu7juaads0pprjprkqww7u7dqrf.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6100012c74246816e9689d98

**Comment by nechai** [comment]
Made clean up commits in this PR. Updated with `--force `push.

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2300-uwcurgu7juaads0pprjprkqww7u7dqrf.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6100012c74246816e9689d98

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2300-uwcurgu7juaads0pprjprkqww7u7dqrf.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6100012c74246816e9689d98

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2300-uwcurgu7juaads0pprjprkqww7u7dqrf.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6100012c74246816e9689d98

**Review by ribel — APPROVED** [review]

**Comment by ribel** [comment]
@ronaldtebrake, @navneet0693 We finally have all green here and ready to be merged as for me.
Can we do this? Then it will also unblock PR #2277, which are one of the biggest left in the 10.3 milestone.

File: modules/social_features/social_group/modules/social_group_flexible_group/social_group_flexible_group.install
Comment (ribel) [reply modules/social_features/social_group/modules/social_group_flexible_group/social_group_flexible_group.install:418]: As I understand here we always install new configs, so even if they will be changed later it should be correct to the instal folder, so we don't need to duplicate it in static. Or are there other reasons?

Comment (ribel) [reply modules/social_features/social_group/modules/social_group_flexible_group/social_group_flexible_group.install:418]: @nechai please use file copies from the `config/static` folder in update hooks. You can check this PR as an example: #2277

**Review by ribel — CHANGES_REQUESTED** [review]
@nechai We need to fix the last thing here (config/static) and after that, this PR can be merged.

File: modules/social_features/social_activity/social_activity.tokens.inc
Comment (ribel) [inline modules/social_features/social_activity/social_activity.tokens.inc:132]: @nechaiCustomer who requested this feature asked to rephrase it with `+ @count other groups` instead of `+ @count groups more` and I think it makes sense to change it.

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2300-uwcurgu7juaads0pprjprkqww7u7dqrf.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6100012c74246816e9689d98

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2300-uwcurgu7juaads0pprjprkqww7u7dqrf.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6100012c74246816e9689d98

**Comment by navneet0693** [comment]
@nechai Please resolve the merge conflicts.

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2300-uwcurgu7juaads0pprjprkqww7u7dqrf.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6100012c74246816e9689d98

**Review by navneet0693 — APPROVED** [review]

**Review by ribel — APPROVED** [review]

**Review by Ressinel — APPROVED** [review]

**Comment by navneet0693** [comment]
🍒  - picked to 10.3.x

**Pull request merged**
